File: src-backend/services/excel-import.js
/**
 * excel-import.js
 * Handles Excel/CSV ingestion with dynamic merge-tag column mapping.
 * Returns structured { validRecords, errorLogs } for clean API responses.
 */

const ExcelJS = require('exceljs');
const fs = require('fs');
const path = require('path');
const { openDatabase } = require('../db/client');
const { normalizePhone, normalizeBatch, sanitizeString } = require('./phone-normalizer');

const uploadDir = path.join(__dirname, '..', 'uploads');
if (!fs.existsSync(uploadDir)) {
  fs.mkdirSync(uploadDir, { recursive: true });
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Load an ExcelJS workbook from a file path (xlsx or csv).
 */
const loadWorkbook = async (filePath) => {
  const workbook = new ExcelJS.Workbook();
  if (filePath.toLowerCase().endsWith('.csv')) {
    await workbook.csv.readFile(filePath);
  } else {
    await workbook.xlsx.readFile(filePath);
  }
  return workbook;
};

/**
 * Extract column header names from the first row of the first sheet.
 * @param {string} filePath
 * @returns {Promise<string[]>}
 */
const extractHeaders = async (filePath) => {
  const workbook = await loadWorkbook(filePath);
  const worksheet = workbook.worksheets[0];
  const headerRow = worksheet.getRow(1);
  return headerRow.values
    .slice(1)
    .map((v) => sanitizeString(v))
    .filter(Boolean);
};

/**
 * Read all data rows as a flat array of objects keyed by header name.
 * Empty rows (all blank cells) are skipped.
 * @param {string} filePath
 * @returns {Promise<Array<Record<string, string>>>}
 */
const readRows = async (filePath) => {
  const workbook = await loadWorkbook(filePath);
  const worksheet = workbook.worksheets[0];
  const headers = worksheet.getRow(1).values
    .slice(1)
    .map((v) => sanitizeString(v));
  const rows = [];
  worksheet.eachRow((row, rowNumber) => {
    if (rowNumber === 1) return;
    const values = row.values.slice(1);
    const entry = {};
    headers.forEach((header, idx) => {
      entry[header] = sanitizeString(values[idx]);
    });
    if (Object.values(entry).some((v) => v !== '')) {
      rows.push(entry);
    }
  });
  return rows;
};

// ─── Mapping Schema ────────────────────────────────────────────────────────────
/**
 * mapping = {
 *   phoneColumn: 'Teléfono',        // required
 *   nameColumn: 'Nombre',           // optional
 *   mergeTags: {                     // dynamic — any number of tags
 *     empresa: 'Empresa',
 *     ciudad: 'Ciudad',
 *     producto: 'Producto Interés'
 *   }
 * }
 */

/**
 * Apply the mapping to a raw row, returning a normalized record shape.
 */
const applyMapping = (row, mapping) => {
  const phone = row[mapping.phoneColumn] || '';
  const name = sanitizeString(row[mapping.nameColumn] || '');
  const mergeTags = {};
  if (mapping.mergeTags && typeof mapping.mergeTags === 'object') {
    for (const [tagName, columnName] of Object.entries(mapping.mergeTags)) {
      mergeTags[sanitizeString(tagName)] = sanitizeString(row[columnName] || '');
    }
  }
  return { phone, name, mergeTags };
};

// ─── Public API ───────────────────────────────────────────────────────────────

/**
 * Preview an import: parse rows, apply mapping, validate phones.
 * Returns { validRecords[], errorLogs[], totalRows }.
 *
 * @param {string} filePath
 * @param {object} mapping
 * @param {string} [defaultCountry='+57']
 * @returns {Promise<{ validRecords: object[], errorLogs: object[], totalRows: number }>}
 */
const previewImport = async (filePath, mapping, defaultCountry = '+57') => {
  const rows = await readRows(filePath);
  const mapped = rows.map((row) => applyMapping(row, mapping));

  const seen = new Set();
  const validRecords = [];
  const errorLogs = [];

  mapped.forEach((record, index) => {
    const rowIndex = index + 1;
    const result = normalizePhone(record.phone, defaultCountry);

    if (!result.valid) {
      errorLogs.push({
        row: rowIndex,
        originalPhone: record.phone,
        originalRaw: result.originalRaw,
        name: record.name,
        error: result.error
      });
      return;
    }

    if (seen.has(result.phone)) {
      errorLogs.push({
        row: rowIndex,
        originalPhone: record.phone,
        originalRaw: result.originalRaw,
        name: record.name,
        error: 'Número duplicado en este archivo'
      });
      return;
    }

    seen.add(result.phone);
    validRecords.push({
      rowIndex,
      normalizedPhone: result.phone,
      name: record.name,
      mergeTags: record.mergeTags
    });
  });

  return { validRecords, errorLogs, totalRows: rows.length };
};

/**
 * Commit validated records into the contacts table.
 * Stores merge tags in custom_vars_json column.
 * Ensures the column exists before writing (additive migration).
 *
 * @param {string} filePath
 * @param {object} mapping
 * @param {string} [defaultCountry='+57']
 * @param {string} [source='excel']
 * @returns {Promise<{ totalRows: number, imported: number, rejected: number, errorLogs: object[] }>}
 */
const commitImport = async (filePath, mapping, defaultCountry = '+57', source = 'excel') => {
  const db = openDatabase();

  // Additive migration: ensure custom_vars_json column exists
  try {
    await db.run('ALTER TABLE contacts ADD COLUMN custom_vars_json TEXT');
  } catch (_) {
    // Column already exists — safe to ignore
  }

  const { validRecords, errorLogs, totalRows } = await previewImport(filePath, mapping, defaultCountry);

  const insertSql = `
    INSERT OR IGNORE INTO contacts
      (phone, country_code, name, var1, var2, custom_vars_json, source, updated_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'))
  `;

  for (const record of validRecords) {
    const var1 = record.mergeTags?.var1 || '';
    const var2 = record.mergeTags?.var2 || '';
    const customVarsJson = JSON.stringify(record.mergeTags || {});
    await db.run(insertSql, [
      record.normalizedPhone,
      defaultCountry,
      record.name,
      var1,
      var2,
      customVarsJson,
      source
    ]);
  }

  // Log the import job
  const jobSql = `
    INSERT INTO import_jobs
      (filename, mapping_json, rows_total, rows_valid, rows_invalid, duplicates_count, errors_json)
    VALUES (?, ?, ?, ?, ?, ?, ?)
  `;
  const duplicatesCount = errorLogs.filter((e) => e.error.includes('duplicado') || e.error.includes('Duplicado')).length;
  await db.run(jobSql, [
    path.basename(filePath),
    JSON.stringify(mapping),
    totalRows,
    validRecords.length,
    errorLogs.length,
    duplicatesCount,
    JSON.stringify(errorLogs)
  ]);

  await db.close();

  return {
    totalRows,
    imported: validRecords.length,
    rejected: errorLogs.length,
    errorLogs
  };
};

module.exports = { extractHeaders, previewImport, commitImport, uploadDir };

File: src-backend/services/excel_import.py
import os
import json
import datetime
import pandas as pd
from typing import List, Dict, Any, Tuple
from db.client import db_client
from services.phone_normalizer import normalize_phone, sanitize_string

UPLOAD_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "uploads"))
os.makedirs(UPLOAD_DIR, exist_ok=True)

def extract_headers(file_path: str) -> List[str]:
    """
    Lee solo los encabezados del archivo Excel o CSV de forma rápida.
    """
    try:
        if file_path.lower().endswith(".csv"):
            df = pd.read_csv(file_path, nrows=0)
        else:
            df = pd.read_excel(file_path, nrows=0)
        
        headers = [sanitize_string(col) for col in df.columns]
        return [h for h in headers if h]
    except Exception as e:
        raise Exception(f"Error leyendo encabezados del archivo: {str(e)}")

def parse_row_with_mapping(row: Dict[str, Any], mapping: Dict[str, Any]) -> Dict[str, Any]:
    """
    Extrae campos mapeados de una fila de datos según el mapeo de columnas.
    """
    phone_col = mapping.get("phoneColumn")
    name_col = mapping.get("nameColumn")
    
    phone = str(row.get(phone_col)) if row.get(phone_col) is not None else ""
    name = sanitize_string(row.get(name_col)) if name_col and row.get(name_col) is not None else ""
    
    merge_tags = {}
    merge_tags_mapping = mapping.get("mergeTags", {})
    if isinstance(merge_tags_mapping, dict):
        for tag_name, col_name in merge_tags_mapping.items():
            val = row.get(col_name)
            merge_tags[sanitize_string(tag_name)] = sanitize_string(val) if val is not None else ""
            
    return {
        "phone": phone,
        "name": name,
        "merge_tags": merge_tags
    }

async def preview_import(file_path: str, mapping: Dict[str, Any], default_country: str = "+57") -> Dict[str, Any]:
    """
    Previsualiza una importación de contactos validando números de teléfono y duplicados en fragmentos.
    """
    valid_records = []
    error_logs = []
    seen = set()
    total_rows = 0
    row_index = 0
    
    is_csv = file_path.lower().endswith(".csv")
    
    # Procesar en fragmentos de 1000 filas para evitar picos de memoria RAM (OOM protection)
    chunk_iterator = pd.read_csv(file_path, chunksize=1000) if is_csv else pd.read_excel(file_path, chunksize=1000)
    
    for chunk in chunk_iterator:
        # Convertir NaNs de Pandas a None para procesar cadenas sin errores
        chunk = chunk.where(pd.notnull(chunk), None)
        
        for _, row in chunk.iterrows():
            row_index += 1
            row_dict = row.to_dict()
            
            # Ignorar filas totalmente vacías
            if not any(v is not None and str(v).strip() != "" for v in row_dict.values()):
                continue
            
            total_rows += 1
            record = parse_row_with_mapping(row_dict, mapping)
            norm = normalize_phone(record["phone"], default_country)
            
            if not norm["valid"]:
                error_logs.append({
                    "row": row_index + 1,  # Fila 1 es el encabezado
                    "originalPhone": record["phone"],
                    "originalRaw": norm.get("originalRaw", ""),
                    "name": record["name"],
                    "error": norm["error"]
                })
                continue
                
            norm_phone = norm["phone"]
            if norm_phone in seen:
                error_logs.append({
                    "row": row_index + 1,
                    "originalPhone": record["phone"],
                    "originalRaw": norm.get("originalRaw", ""),
                    "name": record["name"],
                    "error": "Número duplicado en este archivo"
                })
                continue
                
            seen.add(norm_phone)
            valid_records.append({
                "rowIndex": row_index + 1,
                "normalizedPhone": norm_phone,
                "name": record["name"],
                "mergeTags": record["merge_tags"]
            })
            
    return {
        "validRecords": valid_records,
        "errorLogs": error_logs,
        "totalRows": total_rows
    }

async def commit_import(file_path: str, mapping: Dict[str, Any], default_country: str = "+57", source: str = "excel") -> Dict[str, Any]:
    """
    Confirma la importación escribiendo los contactos válidos en lote dentro de SQLite.
    """
    # Agregar columna custom_vars_json de manera segura si no existía (Migración Aditiva)
    try:
        await db_client.run("ALTER TABLE contacts ADD COLUMN custom_vars_json TEXT")
    except Exception:
        pass # Columna ya existente
        
    # Obtener leads validados
    preview = await preview_import(file_path, mapping, default_country)
    valid_records = preview["validRecords"]
    error_logs = preview["errorLogs"]
    total_rows = preview["totalRows"]

    insert_sql = """
        INSERT OR IGNORE INTO contacts
        (phone, country_code, name, var1, var2, custom_vars_json, source, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'))
    """

    # Generar tuplas de inserción por lotes
    params_list = []
    for r in valid_records:
        tags = r.get("mergeTags", {})
        var1 = tags.get("var1") or tags.get("empresa") or ""
        var2 = tags.get("var2") or tags.get("ciudad") or ""
        custom_vars_json = json.dumps(tags)
        params_list.append((
            r["normalizedPhone"],
            default_country,
            r["name"],
            var1,
            var2,
            custom_vars_json,
            source
        ))

    # Inserción en lote (Batch Write) no bloqueante
    if params_list:
        await db_client.executemany(insert_sql, params_list)

    # Registrar el Import Job
    job_sql = """
        INSERT INTO import_jobs
        (filename, mapping_json, rows_total, rows_valid, rows_invalid, duplicates_count, errors_json, created_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, datetime('now'))
    """
    filename = os.path.basename(file_path)
    duplicates_count = sum(1 for e in error_logs if "duplicado" in e.get("error", "").lower())
    
    await db_client.run(job_sql, (
        filename,
        json.dumps(mapping),
        total_rows,
        len(valid_records),
        len(error_logs),
        duplicates_count,
        json.dumps(error_logs)
    ))

    return {
        "totalRows": total_rows,
        "imported": len(valid_records),
        "rejected": len(error_logs),
        "errorLogs": error_logs
    }
